### PR TITLE
[VM] Added acquires list to file format

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- In upcoming PRs, it will be required for functions to annotate their global storage accesses
- This is the list of resources that they acquire from BorrowGlobal, MoveFrom, and transitively from other functions in the same module
- This adds the list of acquired resources to `StructDefinition`
- The resources in the list do not have type arguments as that is not yet supported by global storage
- The list is not in FunctionSignature as it is not needed outside of the module (much like public/native flags are not needed outside)

## Test Plan

cargo check; cargo test

Cannot properly test until we have parsing in the IR and the bytecode verifier running

### CODE REVIEW

File: language/vm/src/file_format.rs [hunk 1]
```diff
@@ -328,6 +328,14 @@ pub struct FunctionDefinition {
     pub function: FunctionHandleIndex,
     /// Flags for this function (private, public, native, etc.)
     pub flags: u8,
+    /// List of nominal resources (local to this module) that the procedure might access
+    /// Either through: BorrowGlobal, MoveFrom, or transitively through another procedure
+    ///
+    /// Not in the signature as it is not needed outside of the declare modules
```
Comment (shazqadeer) [inline language/vm/src/file_format.rs]: declare modules -> declaring module

File: language/vm/src/file_format.rs [hunk 2]
```diff
@@ -328,6 +328,14 @@ pub struct FunctionDefinition {
...
+    ///
+    /// Note, there is no LocalsSignatureIndex with each struct definition index, as global
```
Comment (shazqadeer) [inline language/vm/src/file_format.rs]: What is the connection between LocalsSignatureIndex and type arguments?

File: language/vm/src/serializer.rs [hunk 3]
```diff
@@ -393,9 +393,30 @@ fn serialize_function_definition(
 ) -> Result<()> {
     write_u16_as_uleb128(binary, function_definition.function.0)?;
     binary.push(function_definition.flags)?;
+    serialize_struct_definition_indicies(binary, &function_definition.acquires_global_resources)?;
     serialize_code_unit(binary, &function_definition.code)
 }
 
+/// Serializes a `Vec<StructDefinitionIndex>`.
+fn serialize_struct_definition_indicies(
```
Comment (shazqadeer) [inline language/vm/src/serializer.rs]: indicies -> indices

File: language/vm/src/deserializer.rs [hunk 4]
```diff
@@ -809,14 +809,28 @@ fn load_function_def(cursor: &mut Cursor<&[u8]>) -> BinaryLoaderResult<FunctionD
...
     Ok(FunctionDefinition {
         function: FunctionHandleIndex(function),
         flags,
+        acquires_global_resources,
         code: code_unit,
     })
 }
 
+/// Deserializes a `Vec<StructDefinitionIndex>`.
+fn load_struct_definition_indicies(
```
[the remark this answers is not in the document]
Comment (shazqadeer) [reply language/vm/src/deserializer.rs]: Please remind me how serde is tested today.  Is the functional tests infrastructure that @vgao1996 worked on somehow also testing serde?

[the remark this answers is not in the document]
Comment (tnowacki) [reply language/vm/src/deserializer.rs]: I think its mostly tested with proptests

File: language/vm/src/file_format.rs (hunk 2 above)
Comment (tnowacki) [reply language/vm/src/file_format.rs]: It should really be renamed. `LocalsSignatureIndex` is used by generics as a way of storing a vector of signature tokens in a pool

File: language/vm/src/deserializer.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/vm/src/deserializer.rs]: For proptests and the core modules, we serialize the generated code and deserialize it and then verify it did not change

**Pull request merged**